File: apps/api/src/services/monkey/hindsightRegret.ts
/**
 * hindsightRegret.ts — legibility-gated counterfactual PREDICTION ERROR.
 *
 * DESIGN HYPOTHESIS (operator-approved redesign of PR #1038, 2026-05-29).
 * Built cleanly, flag-gated OFF (MONKEY_HINDSIGHT_REGRET_LIVE), for operator
 * review — NOT a finished truth. Replaces the rejected v1 (fixed 30-min
 * window + fixed dopamine caps + best-favourable-excursion target +
 * dopamine-only "pain") which was a knob dressed as chemistry.
 *
 * ── WHAT CHANGED FROM v1 (all four rejected mechanisms removed) ──────────
 *   1. NOT fixed pain. The signal is legibility-gated counterfactual
 *      PREDICTION ERROR, scaled by the kernel's OWN outcome distribution
 *      (median+MAD z-score — the same observer scale observerFibCoefficient
 *      / pushReward use). No 0.5 / 0.3 / 0.05 taste constants.
 *   2. WINDOW is the kernel's own qig-warp / foresight expectation HORIZON
 *      (passed in at registration), not 30 min. Fallback horizon is derived
 *      from observed regime persistence, documented at the call site.
 *   3. Counterfactual target = pnl at END OF THE DERIVED HORIZON (and/or
 *      bounded by an observed regime flip), NOT the max favourable excursion.
 *   4. Full NT VECTOR (not dopamine-only): dopamine (signed prediction
 *      error), ACh (bind the cues), NE (salience), serotonin (patience /
 *      temporal-trust), GABA (TARGETED to the same premature-close pattern,
 *      never global), endorphin (relief on a good close).
 *
 * ── PURITY KEYSTONE: the eligibility / legibility gate ───────────────────
 * Regret is scaled by how strongly ALL THREE hold (see `legibilityStrength`):
 *   (1) the kernel OWNED the close (kernel-initiated; operator / manual /
 *       liquidation closes are NOT self-regret),
 *   (2) the continuation was LEGIBLE at close time — the close-time sense
 *       bundle showed evidence to HOLD (qig-warp expectation favoured the
 *       held side, basin direction still with the position, trend coherent),
 *   (3) the SAME REGIME persisted through the derived horizon.
 * If continuation was NOT legible → the move is SURPRISE / NOISE, not regret
 * → NO aversive signal. Weakly legible continuations remain learnable but
 * weak: the observer-scaled prediction-error salience is multiplied by the
 * close-time legibility strength. This maps directly onto QIG canon §31
 * (Sensory Intake & Predictive Coding): a prediction error you could NOT have
 * predicted is surprise, not a learnable mistake. Regret = a LEGIBLE
 * prediction error (low surprise at close, continuation foreseeable).
 *
 * ── QIG canon alignment (vex/kernel/consciousness/neurochemistry.py §29) ─
 *   The six-chemical model = six Cartan generators of E6:
 *     ACh  (ENTRAIN/E1)  — intake/consolidation gate → bind the cues
 *     dop  (AMPLIFY/E2)  — reward signal / prediction error
 *     GABA (DAMPEN/E3)   — inhibition (here: TARGETED, not global)
 *     ser  (ROTATE/E4)   — stability / patience / temporal trust
 *     NE   (NUCLEATE/E5) — salience / surprise magnitude (observer-scaled)
 *     endo (DISSOLVE/E6) — convergence reward / relief
 *   We do NOT invent chemistry semantics; the signs below follow canon's
 *   role of each chemical.
 *
 * DOCTRINAL ANCHORS:
 *   - P1 (Observer sets all params): every magnitude is normalised against
 *     the kernel's own realised pnl_frac distribution (median+MAD). No
 *     hardcoded sting amplitude, no caps chosen by intuition.
 *   - P14 (Variable Separation): hindsight is its OWN chemistry channel,
 *     not folded into the realised-pnl reward.
 *   - P15 (Fail-Closed Safety): invalid price/margin/qty/side/history →
 *     NO chemistry mutation (zero vector). Never blocks trading.
 *
 * This module is PURE (no I/O, no time, no DB). Orchestration (register a
 * watch at close with the sense bundle + horizon, advance with live price,
 * resolve at horizon end) lives in loop.ts behind the flag.
 */

// ── E6 six-chemical NT delta vector (canon §29.1) ────────────────────────
// Signs are the chemistry's CANONICAL ROLE, not free parameters. Magnitudes
// are observer-scaled (see deriveMagnitude) and, on regret, multiplied by
// close-time legibility. Equal per-channel salience remains a dark-mode draft
// hypothesis until a channel-envelope derivation is approved. Flag-OFF callers
// never read it.
export interface HindsightNtDeltas {
  /** Signed prediction error. <0 on a legible premature close (the trade
   *  was right, the exit wrong); small >0 relief on a good close. */
  dopamineDelta: number;
  /** ACh ↑ — bind the close-time cues for next time (intake/consolidation
   *  gate). Non-negative; salient on both regret and good-close. */
  acetylcholineDelta: number;
  /** NE ↑ — salience of the event, observer-scaled. Non-negative. */
  norepinephrineDelta: number;
  /** serotonin — patience / temporal-trust. <0 on premature close (correct
   *  the impatience), >0 on a good close (timing confidence). */
  serotoninDelta: number;
  /** GABA ↑ — TARGETED inhibition keyed to the premature-close PATTERN
   *  (regime + side), NOT global suppression. Non-negative; only on regret.
   *  Carries `gabaTarget` so the kernel binds it to a pattern, not to "close
   *  less often". */
  gabaDelta: number;
  /** endorphin — relief on a good close (avoided a worse counterfactual).
   *  Non-negative; zero on regret. */
  endorphinDelta: number;
}

/** A pattern key the targeted-GABA delta is bound to (regime|side), so the
 *  kernel learns "don't cut THIS kind of position early" rather than "fear
 *  closing". null when not a regret (no target to bind). */
export type GabaTarget = string | null;

/** The close-time SENSE BUNDLE — captured at the moment of a kernel close.
 *  Drives the legibility gate (§31 predictive coding) and the targeted-GABA
 *  binding. Every field is a sense already available in SymbolState / the
 *  qig-warp expectation at the close site (see loop.ts mapping). */
export interface CloseSenseBundle {
  // ── ownership (gate condition 1) ──
  /** True iff the kernel initiated this close (own exit policy fired).
   *  Operator/manual/liquidation closes are false → routed to exemplar
   *  learning, never self-regret. */
  kernelOwnedClose: boolean;

  // ── market senses at close ──
  /** Held side as a sign: +1 long, -1 short. */
  sideSign: 1 | -1;
  /** qig-warp expectation direction at close (-1 short-favoured, 0 flat/
   *  observe, +1 long-favoured). From ExpectationDecision.expectation_direction. */
  warpExpectationSign: -1 | 0 | 1;
  /** qig-warp expectation confidence at close [0,1].
   *  From ExpectationDecision.expectation_confidence. */
  warpExpectationConfidence: number;
  /** qig-warp regime label at close (e.g. 'aligned'|'reverse_tape'|'chop'). */
  regimeAtClose: string;
  /** basinDir at close (signed market direction; +long/-short lean). */
  basinDirAtClose: number;
  /** tape trend at close (signed). */
  tapeTrendAtClose: number;
  /** coherenceStreak at close — consecutive ticks the kernel's own
   *  perception+forecast stayed coherent on the held position. Higher =
   *  the hold was more legible. */
  coherenceStreak: number;
}

/** The counterfactual outcome resolved at the END of the derived horizon. */
export interface CounterfactualOutcome {
  /** Realised net pnl (USDT) the close actually booked (Polo-authoritative). */
  realizedPnlUsdt: number;
  /** Counterfactual pnl (USDT) of having HELD to the end of the horizon
   *  (or to the observed regime flip, whichever bounded first). NOT max
   *  favourable excursion. */
  horizonEndPnlUsdt: number;
  /** Margin (USDT) backing the closed position — for pnl_frac normalisation. */
  marginUsdt: number;
  /** True iff the SAME regime that was legible at close persisted through
   *  the horizon (gate condition 3). If the regime flipped to an unrelated
   *  move, the continuation is not the one the kernel could have foreseen. */
  regimePersisted: boolean;
}

/** Output of the resolve transform — the NT vector, its GABA target, and
 *  telemetry. */
export interface HindsightResult {
  nt: HindsightNtDeltas;
  gabaTarget: GabaTarget;
  /** Foregone gain (USDT): max(0, horizonEnd - realized). 0 on a good close. */
  foregoneGainUsdt: number;
  /** The observer-scaled prediction-error magnitude actually consumed
   *  (z-scored regret fraction). */
  predictionErrorZ: number;
  /** Which branch fired:
   *    'hindsight_regret'      — eligible + holding would have won
   *    'hindsight_good_close'  — eligible + close avoided a worse outcome
   *    'ineligible_noise'      — gate failed (surprise/noise) → zero vector
   *    'ineligible_not_owned'  — not a kernel close → routed elsewhere
   *    'hindsight_invalid' / 'hindsight_no_margin' — fail-closed */
  source: string;
}

const EPS = 1e-12;
/** Minimum samples before the observer scale is trusted (mirrors pushReward
 *  / observer_fib_coefficient). Below this we fall through with no signal —
 *  cold-start must NOT emit chemistry from an unestablished scale. */
const MIN_SAMPLES = 5;

function isFiniteNumber(x: unknown): x is number {
  return typeof x === 'number' && Number.isFinite(x);
}

const ZERO_NT: HindsightNtDeltas = Object.freeze({
  dopamineDelta: 0,
  acetylcholineDelta: 0,
  norepinephrineDelta: 0,
  serotoninDelta: 0,
  gabaDelta: 0,
  endorphinDelta: 0,
});

function ineligible(source: string): HindsightResult {
  return {
    nt: { ...ZERO_NT },
    gabaTarget: null,
    foregoneGainUsdt: 0,
    predictionErrorZ: 0,
    source,
  };
}

/**
 * Median absolute deviation around the median. Robust to outliers — the
 * SAME statistic observerFibCoefficient / pushReward use for the observer
 * scale. Returns the (median, mad) pair so the z-score matches canon.
 */
export function medianAndMad(xs: number[]): { median: number; mad: number } {
  const finite = xs.filter(isFiniteNumber);
  if (finite.length === 0) return { median: 0, mad: 0 };
  const sorted = [...finite].sort((a, b) => a - b);
  const med = (arr: number[]): number =>
    arr.length % 2 === 0
      ? (arr[arr.length / 2 - 1]! + arr[arr.length / 2]!) / 2
      : arr[Math.floor(arr.length / 2)]!;
  const median = med(sorted);
  const devs = sorted.map((x) => Math.abs(x - median)).sort((a, b) => a - b);
  return { median, mad: med(devs) };
}

/**
 * Counterfactual pnl (USDT) of holding the CLOSED position to `price`.
 *   long  : (price - exitPrice) * qty   — gains if price rises after exit
 *   short : (exitPrice - price) * qty   — gains if price falls after exit
 * Added to realised pnl → total "if I'd just left it" pnl. null on invalid.
 */
export function counterfactualPnlUsdt(
  args: { sideSign: 1 | -1; qty: number; exitPrice: number; realizedPnlUsdt: number },
  price: number,
): number | null {
  if (!isFiniteNumber(price) || price <= 0) return null;
  if (!isFiniteNumber(args.qty) || args.qty <= 0) return null;
  if (!isFiniteNumber(args.exitPrice) || args.exitPrice <= 0) return null;
  if (!isFiniteNumber(args.realizedPnlUsdt)) return null;
  if (args.sideSign !== 1 && args.sideSign !== -1) return null;
  const marginal = (price - args.exitPrice) * args.sideSign * args.qty;
  return args.realizedPnlUsdt + marginal;
}

/**
 * PURITY KEYSTONE — the eligibility / legibility gate.
 *
 * Returns true iff regret may fire. The actual regret magnitude is multiplied
 * by `legibilityStrength`; weak evidence cannot produce full-strength regret.
 * All three conditions:
 *   (1) kernel owned the close,
 *   (2) the continuation was LEGIBLE at close — the close-time senses showed
 *       evidence to HOLD: qig-warp expectation favoured the held side with
 *       confidence, basin direction still leaned with the position, and the
 *       hold was coherent (coherenceStreak > 0). "Legible" = the kernel's
 *       OWN forecast pointed at continuation (low surprise per §31), so a
 *       continuation is a prediction error it could have avoided.
 *   (3) the same regime persisted through the horizon.
 *
 * Legibility uses the kernel's own observer-relative signal, NOT a fixed
 * threshold: the warp expectation must be ALIGNED with the held side and
 * carry positive confidence, AND basinDir must still lean the held way. The
 * strength is continuous: confidence × basin-alignment magnitude × coherence
 * strength × regime-persistence.
 *
 * If legibility fails → SURPRISE / NOISE → no aversive signal (canon §31:
 * unpredictable continuation is surprise, not a learnable mistake).
 */
function clamp01(x: number): number {
  if (!isFiniteNumber(x)) return 0;
  return Math.max(0, Math.min(1, x));
}

/** Continuous legibility multiplier ∈ [0,1]. Weak confidence / tiny basin
 *  lean / one-tick coherence scales regret down instead of merely opening a
 *  Boolean gate. Regime persistence is currently observed as a latch, so its
 *  strength is 1 when persisted and 0 after a flip. */
export function legibilityStrength(b: CloseSenseBundle, regimePersisted = true): number {
  if (!regimePersisted) return 0;
  // qig-warp expectation must have favoured the HELD side (continuation),
  // i.e. expectation direction agrees with sideSign, with > 0 confidence.
  const warpStrength =
    b.warpExpectationSign === b.sideSign
      ? clamp01(b.warpExpectationConfidence)
      : 0;
  // basin direction must still lean the held way (sign agreement).
  const basinStrength =
    isFiniteNumber(b.basinDirAtClose) &&
    Math.sign(b.basinDirAtClose) === b.sideSign
      ? Math.tanh(Math.abs(b.basinDirAtClose))
      : 0;
  const coherence =
    isFiniteNumber(b.coherenceStreak) && b.coherenceStreak > 0
      ? b.coherenceStreak / (b.coherenceStreak + 1)
      : 0;
  return clamp01(warpStrength * basinStrength * coherence);
}

export function isContinuationLegible(b: CloseSenseBundle): boolean {
  return legibilityStrength(b) > EPS;
}

/** Full eligibility = owned ∧ legible-at-close ∧ regime-persisted. */
export function isEligibleForRegret(
  b: CloseSenseBundle,
  regimePersisted: boolean,
): { eligible: boolean; reason: 'eligible' | 'not_owned' | 'not_legible' | 'regime_changed' } {
  if (!b.kernelOwnedClose) return { eligible: false, reason: 'not_owned' };
  if (!isContinuationLegible(b)) return { eligible: false, reason: 'not_legible' };
  if (!regimePersisted) return { eligible: false, reason: 'regime_changed' };
  return { eligible: true, reason: 'eligible' };
}

/**
 * Observer-scaled prediction-error magnitude. `frac` is foregoneGain/margin
 * (a pnl-fraction). z-scored against the kernel's own pnl_frac distribution
 * (median+MAD), exactly like observerFibCoefficient. Returns the z-magnitude
 * (>= 0) and the salience weight tanh(z) ∈ [0,1) used to scale every NT
 * channel. With < MIN_SAMPLES or zero MAD → returns null (no trusted scale
 * → caller emits nothing; cold-start must not fabricate chemistry).
 */
export function deriveMagnitude(
  frac: number,
  pnlFracHistory: number[],
): { z: number; salience: number } | null {
  if (!isFiniteNumber(frac)) return null;
  if (pnlFracHistory.length < MIN_SAMPLES) return null;
  const { mad } = medianAndMad(pnlFracHistory);
  if (mad <= EPS) return null;
  const z = Math.abs(frac) / mad;
  return { z, salience: Math.tanh(z) };
}

/** GABA target key: bind targeted inhibition to the (regime, side) pattern
 *  of the premature close so the kernel learns "don't cut THIS kind of
 *  position early" — NOT "close less". */
export function gabaTargetKey(b: CloseSenseBundle): string {
  const side = b.sideSign === 1 ? 'long' : 'short';
  const regime = typeof b.regimeAtClose === 'string' && b.regimeAtClose.length > 0
    ? b.regimeAtClose
    : 'unknown';
  return `premature_close:${regime}:${side}`;
}

/**
 * Resolve a watch at horizon end into the full NT vector. PURE.
 *
 * Branches:
 *   - fail-closed (invalid pnl/margin) → zero vector.
 *   - not owned → 'ineligible_not_owned', zero vector (route to exemplar
 *     learning is the caller's job; this signal emits nothing).
 *   - foregoneGain <= 0 (close avoided a worse outcome) AND eligible-by-
 *     ownership → 'hindsight_good_close': relief vector (dop small+, ACh+,
 *     ser+, endo+, NE if salient, GABA 0).
 *   - foregoneGain > 0 but NOT legible / regime changed → 'ineligible_noise',
 *     zero vector (the continuation was surprise, not a foreseeable mistake).
 *   - foregoneGain > 0 AND eligible → 'hindsight_regret': aversive vector
 *     (dop-, ACh+, NE+, ser-, GABA+ TARGETED, endo 0), all observer-scaled.
 */
export function resolveHindsight(
  bundle: CloseSenseBundle,
  outcome: CounterfactualOutcome,
  pnlFracHistory: number[],
): HindsightResult {
  const { realizedPnlUsdt: realized, horizonEndPnlUsdt: horizonEnd, marginUsdt } = outcome;

  // ── fail-closed ───────────────────────────────────────────────────────
  if (!isFiniteNumber(realized) || !isFiniteNumber(horizonEnd)) {
    return ineligible('hindsight_invalid');
  }
  if (!isFiniteNumber(marginUsdt) || marginUsdt <= 0) {
    return ineligible('hindsight_no_margin');
  }

  // ── ownership gate (condition 1) ────────────────────────────────────────
  // A non-owned close is not self-regret; emit nothing here (caller routes
  // to exemplar learning).
  if (!bundle.kernelOwnedClose) {
    return ineligible('ineligible_not_owned');
  }

  const foregoneGainUsdt = horizonEnd - realized;

  // ── good-close branch (close avoided a worse / equal outcome) ───────────
  // Holding would NOT have beaten the close. Relief, not regret. We still
  // require an established observer scale so the magnitude is meaningful.
  if (foregoneGainUsdt <= EPS) {
    const avoidedLossFrac = Math.abs(foregoneGainUsdt) / marginUsdt;
    const mag = deriveMagnitude(avoidedLossFrac, pnlFracHistory);
    if (mag === null) return ineligible('ineligible_noise'); // no trusted scale
    const s = mag.salience;
    // Relief vector (canon roles): dopamine small + (avoided counterfactual
    // loss, observer-scaled — NOT a fixed +0.05), ACh + (bind the good
    // timing), serotonin + (timing confidence / temporal trust), endorphin +
    // (relief / convergence), NE + only if salient, GABA 0 (no inhibition to
    // bind — closing was correct).
    return {
      nt: {
        dopamineDelta: s,
        acetylcholineDelta: s,
        norepinephrineDelta: s,
        serotoninDelta: s,
        gabaDelta: 0,
        endorphinDelta: s,
      },
      gabaTarget: null,
      foregoneGainUsdt: 0,
      predictionErrorZ: mag.z,
      source: 'hindsight_good_close',
    };
  }

  // ── legibility + regime-persistence gate (conditions 2 & 3) ─────────────
  // foregoneGain > 0: holding WOULD have won. But only counts as regret if
  // the continuation was LEGIBLE at close AND the same regime persisted.
  // Otherwise it was surprise/noise — emit nothing.
  const gate = isEligibleForRegret(bundle, outcome.regimePersisted);
  if (!gate.eligible) {
    return ineligible('ineligible_noise');
  }

  // ── regret branch: legible premature close, observer-scaled ─────────────
  const regretFrac = foregoneGainUsdt / marginUsdt;
  const mag = deriveMagnitude(regretFrac, pnlFracHistory);
  if (mag === null) return ineligible('ineligible_noise'); // no trusted scale
  const legibility = legibilityStrength(bundle, outcome.regimePersisted);
  if (legibility <= EPS) return ineligible('ineligible_noise');
  const s = mag.salience * legibility; // ∈ (0,1), observer-scale × legibility

  // Aversive vector. Signs follow canon §29.1 roles; magnitudes = salience
  // (observer-scaled). dopamine NEGATIVE (signed prediction error — the
  // trade was right, the exit wrong); ACh + (bind the close-time cues);
  // NE + (salience of the surprise-avoidable error); serotonin NEGATIVE
  // (patience/temporal-trust correction — the kernel was too impatient);
  // GABA + TARGETED to the (regime,side) pattern (NOT global — see
  // gabaTarget); endorphin 0 (no relief — this was a mistake).
  return {
    nt: {
      dopamineDelta: -s,
      acetylcholineDelta: s,
      norepinephrineDelta: s,
      serotoninDelta: -s,
      gabaDelta: s,
      endorphinDelta: 0,
    },
    gabaTarget: gabaTargetKey(bundle),
    foregoneGainUsdt,
    predictionErrorZ: mag.z,
    source: 'hindsight_regret',
  };
}

/** Feature-flag helper. Default OFF — behaviour byte-identical when unset. */
export function isHindsightRegretLive(): boolean {
  return process.env.MONKEY_HINDSIGHT_REGRET_LIVE === 'true';
}

File: ml-worker/src/monkey_kernel/hindsight_regret.py
"""hindsight_regret.py — legibility-gated counterfactual PREDICTION ERROR.

Python parity for apps/api/src/services/monkey/hindsightRegret.ts. Identical
formula, derivation, source labels, and fail-closed behaviour. See the TS
module's header for the full design rationale; this docstring summarises.

DESIGN HYPOTHESIS (operator-approved redesign of PR #1038, 2026-05-29),
flag-gated OFF (MONKEY_HINDSIGHT_REGRET_LIVE), for operator review — NOT a
finished truth. Replaces the rejected v1 (fixed 30-min window + fixed dopamine
caps + best-favourable-excursion + dopamine-only pain).

The signal is a legibility-gated counterfactual PREDICTION ERROR scaled by the
kernel's OWN outcome distribution (median+MAD z-score — the observer scale
observer_fib_coefficient / push_reward use). No fixed caps/taste constants.

PURITY KEYSTONE — eligibility / legibility gate. Regret is scaled by how
strongly ALL THREE hold:
  (1) the kernel OWNED the close (operator/manual/liquidation → not self-regret),
  (2) the continuation was LEGIBLE at close (qig-warp expectation favoured the
      held side, basin direction still leaned with the position, the hold was
      coherent),
  (3) the same REGIME persisted through the derived horizon.
If continuation was NOT legible → SURPRISE / NOISE, not regret → no aversive
signal. Weakly legible continuations remain learnable but weak: observer-scaled
prediction-error salience is multiplied by close-time legibility strength. Maps
onto QIG canon §31 (Sensory Intake & Predictive Coding): an unpredictable
continuation is surprise, not a learnable mistake.

QIG canon §29.1 (six-chemical E6 Cartan generators) governs the NT signs:
  ACh (ENTRAIN/E1) bind cues · dopamine (AMPLIFY/E2) reward/prediction-error ·
  GABA (DAMPEN/E3) inhibition (TARGETED here) · serotonin (ROTATE/E4)
  patience/temporal-trust · NE (NUCLEATE/E5) salience · endorphin (DISSOLVE/E6)
  relief. We do NOT invent chemistry semantics.

DOCTRINAL ANCHORS: P1 (observer-set magnitudes via median+MAD), P14 (own
channel), P15 (fail-closed — invalid price/margin → zero vector, never blocks).

PURE module (no I/O, no time, no DB).
"""

from __future__ import annotations

import math
import os
from dataclasses import dataclass
from typing import Literal, Optional

_EPS: float = 1e-12
# Minimum samples before the observer scale is trusted (mirrors push_reward /
# observer_fib_coefficient). Below → no signal (cold-start must not fabricate).
_MIN_SAMPLES: int = 5


def _is_finite(x: object) -> bool:
    return isinstance(x, (int, float)) and not isinstance(x, bool) and math.isfinite(float(x))


@dataclass
class HindsightNtDeltas:
    """E6 six-chemical NT delta vector (canon §29.1). Signs are canonical
    roles; magnitudes are observer-scaled. Flag-OFF callers never read it."""

    dopamine_delta: float = 0.0
    acetylcholine_delta: float = 0.0
    norepinephrine_delta: float = 0.0
    serotonin_delta: float = 0.0
    gaba_delta: float = 0.0
    endorphin_delta: float = 0.0


@dataclass
class CloseSenseBundle:
    """Close-time sense bundle — drives the legibility gate + GABA binding."""

    kernel_owned_close: bool
    side_sign: int  # +1 long, -1 short
    warp_expectation_sign: int  # -1 short, 0 flat/observe, +1 long
    warp_expectation_confidence: float
    regime_at_close: str
    basin_dir_at_close: float
    tape_trend_at_close: float
    coherence_streak: int


@dataclass
class CounterfactualOutcome:
    """Counterfactual outcome resolved at the END of the derived horizon."""

    realized_pnl_usdt: float
    horizon_end_pnl_usdt: float  # NOT max favourable excursion
    margin_usdt: float
    regime_persisted: bool


@dataclass
class HindsightResult:
    nt: HindsightNtDeltas
    gaba_target: Optional[str]
    foregone_gain_usdt: float
    prediction_error_z: float
    source: str


def _ineligible(source: str) -> HindsightResult:
    return HindsightResult(
        nt=HindsightNtDeltas(),
        gaba_target=None,
        foregone_gain_usdt=0.0,
        prediction_error_z=0.0,
        source=source,
    )


def median_and_mad(xs: list[float]) -> tuple[float, float]:
    """(median, MAD) — the observer scale observer_fib_coefficient uses."""
    finite = [float(x) for x in xs if _is_finite(x)]
    if not finite:
        return (0.0, 0.0)
    s = sorted(finite)
    n = len(s)

    def _med(arr: list[float]) -> float:
        m = len(arr)
        return (arr[m // 2 - 1] + arr[m // 2]) / 2 if m % 2 == 0 else arr[m // 2]

    median = _med(s)
    devs = sorted(abs(x - median) for x in s)
    return (median, _med(devs))


def counterfactual_pnl_usdt(
    *,
    side_sign: int,
    qty: float,
    exit_price: float,
    realized_pnl_usdt: float,
    price: float,
) -> Optional[float]:
    """Counterfactual pnl (USDT) of holding the CLOSED position to ``price``.

    long  : (price - exit_price) * qty
    short : (exit_price - price) * qty
    Returns realised + marginal. None on any invalid input (fail-closed).
    """
    if not _is_finite(price) or price <= 0:
        return None
    if not _is_finite(qty) or qty <= 0:
        return None
    if not _is_finite(exit_price) or exit_price <= 0:
        return None
    if not _is_finite(realized_pnl_usdt):
        return None
    if side_sign not in (1, -1):
        return None
    marginal = (price - exit_price) * side_sign * qty
    return float(realized_pnl_usdt + marginal)


def _clamp_to_unit_interval(x: float) -> float:
    if not _is_finite(x):
        return 0.0
    return max(0.0, min(1.0, float(x)))


def legibility_strength(b: CloseSenseBundle, regime_persisted: bool = True) -> float:
    """Continuous legibility multiplier ∈ [0,1].

    Weak confidence / tiny basin lean / one-tick coherence scales regret down
    instead of merely opening a Boolean gate. Regime persistence is currently
    observed as a latch, so its strength is 1 when persisted and 0 after a flip.
    """
    if not regime_persisted:
        return 0.0
    warp_strength = (
        _clamp_to_unit_interval(b.warp_expectation_confidence)
        if b.warp_expectation_sign == b.side_sign
        else 0.0
    )
    basin_strength = (
        math.tanh(abs(float(b.basin_dir_at_close)))
        if _is_finite(b.basin_dir_at_close) and _sign(float(b.basin_dir_at_close)) == b.side_sign
        else 0.0
    )
    coherence = (
        float(b.coherence_streak) / float(b.coherence_streak + 1)
        if _is_finite(b.coherence_streak) and b.coherence_streak > 0
        else 0.0
    )
    return _clamp_to_unit_interval(warp_strength * basin_strength * coherence)


def is_continuation_legible(b: CloseSenseBundle) -> bool:
    """PURITY KEYSTONE part — was the continuation foreseeable at close?

    True iff the close-time senses showed evidence to HOLD: qig-warp
    expectation favoured the held side with positive confidence, basin
    direction still leaned the held way, and the hold was coherent for at
    least one tick. "Legible" = the kernel's own forecast pointed at
    continuation (low surprise per §31) → a continuation is a prediction
    error it could have avoided. Otherwise it is surprise/noise.
    """
    return legibility_strength(b) > _EPS


def _sign(x: float) -> int:
    if x > 0:
        return 1
    if x < 0:
        return -1
    return 0


def is_eligible_for_regret(
    b: CloseSenseBundle,
    regime_persisted: bool,
) -> tuple[bool, Literal["eligible", "not_owned", "not_legible", "regime_changed"]]:
    """owned ∧ legible-at-close ∧ regime-persisted."""
    if not b.kernel_owned_close:
        return (False, "not_owned")
    if not is_continuation_legible(b):
        return (False, "not_legible")
    if not regime_persisted:
        return (False, "regime_changed")
    return (True, "eligible")


def derive_magnitude(
    frac: float,
    pnl_frac_history: list[float],
) -> Optional[tuple[float, float]]:
    """(z, salience) observer-scaled magnitude, or None if no trusted scale.

    z = |frac| / MAD(history); salience = tanh(z) ∈ [0,1). With < MIN_SAMPLES
    or MAD≈0 → None (cold-start must not fabricate chemistry).
    """
    if not _is_finite(frac):
        return None
    if len(pnl_frac_history) < _MIN_SAMPLES:
        return None
    _, mad = median_and_mad(pnl_frac_history)
    if mad <= _EPS:
        return None
    z = abs(frac) / mad
    return (z, math.tanh(z))


def gaba_target_key(b: CloseSenseBundle) -> str:
    """Bind targeted GABA to the (regime, side) premature-close pattern."""
    side = "long" if b.side_sign == 1 else "short"
    regime = b.regime_at_close if isinstance(b.regime_at_close, str) and b.regime_at_close else "unknown"
    return f"premature_close:{regime}:{side}"


def resolve_hindsight(
    bundle: CloseSenseBundle,
    outcome: CounterfactualOutcome,
    pnl_frac_history: Optional[list[float]] = None,
) -> HindsightResult:
    """Resolve a watch at horizon end into the full NT vector. PURE.

    Mirrors resolveHindsight() in hindsightRegret.ts exactly.
    """
    if pnl_frac_history is None:
        pnl_frac_history = []

    realized = outcome.realized_pnl_usdt
    horizon_end = outcome.horizon_end_pnl_usdt
    margin = outcome.margin_usdt

    # fail-closed
    if not _is_finite(realized) or not _is_finite(horizon_end):
        return _ineligible("hindsight_invalid")
    if not _is_finite(margin) or margin <= 0:
        return _ineligible("hindsight_no_margin")

    # ownership gate (condition 1)
    if not bundle.kernel_owned_close:
        return _ineligible("ineligible_not_owned")

    foregone_gain = horizon_end - realized

    # good-close branch (avoided a worse/equal outcome)
    if foregone_gain <= _EPS:
        avoided_loss_frac = abs(foregone_gain) / margin
        mag = derive_magnitude(avoided_loss_frac, pnl_frac_history)
        if mag is None:
            return _ineligible("ineligible_noise")
        _z, s = mag
        return HindsightResult(
            nt=HindsightNtDeltas(
                dopamine_delta=float(s),
                acetylcholine_delta=float(s),
                norepinephrine_delta=float(s),
                serotonin_delta=float(s),
                gaba_delta=0.0,
                endorphin_delta=float(s),
            ),
            gaba_target=None,
            foregone_gain_usdt=0.0,
            prediction_error_z=float(_z),
            source="hindsight_good_close",
        )

    # legibility + regime-persistence gate (conditions 2 & 3)
    eligible, _reason = is_eligible_for_regret(bundle, outcome.regime_persisted)
    if not eligible:
        return _ineligible("ineligible_noise")

    # regret branch: legible premature close, observer-scaled
    regret_frac = foregone_gain / margin
    mag = derive_magnitude(regret_frac, pnl_frac_history)
    if mag is None:
        return _ineligible("ineligible_noise")
    z, salience = mag
    legibility = legibility_strength(bundle, outcome.regime_persisted)
    if legibility <= _EPS:
        return _ineligible("ineligible_noise")
    s = salience * legibility

    return HindsightResult(
        nt=HindsightNtDeltas(
            dopamine_delta=float(-s),
            acetylcholine_delta=float(s),
            norepinephrine_delta=float(s),
            serotonin_delta=float(-s),
            gaba_delta=float(s),
            endorphin_delta=0.0,
        ),
        gaba_target=gaba_target_key(bundle),
        foregone_gain_usdt=float(foregone_gain),
        prediction_error_z=float(z),
        source="hindsight_regret",
    )


def is_hindsight_regret_live() -> bool:
    """Feature flag. Default OFF — behaviour byte-identical when unset."""
    return os.environ.get("MONKEY_HINDSIGHT_REGRET_LIVE") == "true"


__all__ = [
    "HindsightNtDeltas",
    "CloseSenseBundle",
    "CounterfactualOutcome",
    "HindsightResult",
    "median_and_mad",
    "counterfactual_pnl_usdt",
    "is_continuation_legible",
    "legibility_strength",
    "is_eligible_for_regret",
    "derive_magnitude",
    "gaba_target_key",
    "resolve_hindsight",
    "is_hindsight_regret_live",
]
